queue.go

```diff
@@ -65,123 +126,291 @@ type Broadcast interface {
+			old.b.Finished()
+			q.deleteItem(old)
 		}
+	} else if !lb.unique {
+		// Slow path, hopefully nothing hot hits this.
+		var remove []*limitedBroadcast
+		q.tq.Ascend(func(item btree.Item) bool {
+			cur := item.(*limitedBroadcast)
+
+			// Special Broadcasts can only invalidate each other.
+			switch cur.b.(type) {
+			case NamedBroadcast:
+				// noop
+			case UniqueBroadcast:
+				// noop
+			default:
+				if b.Invalidates(cur.b) {
+					cur.b.Finished()
+					remove = append(remove, cur)
+				}
+			}
+			return true
+		})
+		for _, cur := range remove {
+			q.deleteItem(cur)
+		}
+	}
+
+	// Append to the relevant queue.
+	q.addItem(lb)
+}
+
+// deleteItem removes the given item from the overall datastructure. You
+// must already hold the mutex.
+func (q *TransmitLimitedQueue) deleteItem(cur *limitedBroadcast) {
+	_ = q.tq.Delete(cur)
+	if cur.name != "" {
+		delete(q.tm, cur.name)
 	}
 
-	// Append to the queue
-	q.bcQueue = append(q.bcQueue, &limitedBroadcast{0, b})
+	if q.tq.Len() == 0 {
+		// At idle there's no reason to let the id generator keep going
+		// indefinitely.
+		q.idGen = 0
+	}
+}
+
+// addItem adds the given item into the overall datastructure. You must already
+// hold the mutex.
+func (q *TransmitLimitedQueue) addItem(cur *limitedBroadcast) {
+	_ = q.tq.ReplaceOrInsert(cur)
+	if cur.name != "" {
+		q.tm[cur.name] = cur
+	}
+}
+
+// getTransmitRange returns a pair of min/max values for transmit values
+// represented by the current queue contents. Both values represent actual
+// transmit values on the interval [0, len). You must already hold the mutex.
+func (q *TransmitLimitedQueue) getTransmitRange() (minTransmit, maxTransmit int) {
+	if q.lenLocked() == 0 {
+		return 0, 0
+	}
+	minItem, maxItem := q.tq.Min(), q.tq.Max()
+	if minItem == nil || maxItem == nil {
+		return 0, 0
+	}
+
+	min := minItem.(*limitedBroadcast).transmits
+	max := maxItem.(*limitedBroadcast).transmits
+
+	return min, max
 }
 
 // GetBroadcasts is used to get a number of broadcasts, up to a byte limit
 // and applying a per-message overhead as provided.
 func (q *TransmitLimitedQueue) GetBroadcasts(overhead, limit int) [][]byte {
-	q.Lock()
-	defer q.Unlock()
+	q.mu.Lock()
+	defer q.mu.Unlock()
 
 	// Fast path the default case
-	if len(q.bcQueue) == 0 {
+	if q.lenLocked() == 0 {
 		return nil
 	}
 
 	transmitLimit := retransmitLimit(q.RetransmitMult, q.NumNodes())
-	bytesUsed := 0
-	var toSend [][]byte
-
-	for i := len(q.bcQueue) - 1; i >= 0; i-- {
-		// Check if this is within our limits
-		b := q.bcQueue[i]
-		msg := b.b.Message()
-		if bytesUsed+overhead+len(msg) > limit {
+
+	var (
+		bytesUsed int
+		toSend    [][]byte
+		reinsert  []*limitedBroadcast
+	)
+
+	// Visit fresher items first, but only look at stuff that will fit.
+	// We'll go tier by tier, grabbing the largest items first.
+	minTr, maxTr := q.getTransmitRange()
+	for transmits := minTr; transmits <= maxTr; /*do not advance automatically*/ {
```

Comment:
Do we need the complexity of the tier first visiting? Can we not just ascend from the Min to the Max in a single walk rather than nested? It seems like it would be effectively the same, by visiting the tiers in order.

Reply:
This is the trickiest part for good reason.  The data structure features that this TransmitLimitedQueue require for normal, efficient operation are varied enough that it was tricky to improve the performance for large queue depths without affecting the other usages:

1. Evict stale messages when superseding messages come in. The map + the two optional interfaces help here.
2. Prune the oldest messages, where oldest mostly means by number of transmits. The btree being bidirectional traversible and primarily sorted by number of transmits helps.
3. Fill a finite knapsack of the Newest and Correctly Sized messages without traversing everything. This is the part that requires the tiered traversal to avoid many degenerate performance scenarios.
 
Any naive traversal of the btree bounded just by the primary index (num transmits) is going to be adversely affected by long runs of incorrectly sized messages, especially as the knapsack gets closer to being full. This can quickly degrade to O(N), which is the thing we're desperately trying to avoid.  An earlier iteration of this was simpler and partly fixed the perf issue, but upon a deeper trace dump of iteration depths I discovered that queue shapes like `[big, big, big, big, big, big, ....., small]` do happen frequently and cause the original bad O(N) traversal behavior to fill the knapsack.

Doing the tiered traversal lets you to only ever visit elements that are _definitely going into the knapsack_ (skipping huge swaths of each tier in the process) and thus avoid any incidental visitations in all situations.